File: src/main/java/com/flexpoker/game/command/events/BaseGameEvent.java
package com.flexpoker.game.command.events;

import java.time.Instant;
import java.util.UUID;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.flexpoker.exception.FlexPokerException;
import com.flexpoker.framework.event.Event;
import com.flexpoker.util.StringUtils;

/**
 * Base class used to handle some standard methods and fields so that the
 * subclasses can just be specific to what they need.
 */
public abstract class BaseGameEvent implements Event {

    private final UUID aggregateId;

    private int version;

    private final Instant time;

    public BaseGameEvent(UUID aggregateId) {
        this.aggregateId = aggregateId;
        this.time = Instant.now();
    }

    @JsonProperty
    @Override
    public UUID getAggregateId() {
        return aggregateId;
    }

    @JsonProperty
    @Override
    public int getVersion() {
        if (version == 0) {
            throw new FlexPokerException("should be calling getVersion() in situations where it's already been set");
        }
        return version;
    }

    @Override
    public void setVersion(int version) {
        this.version = version;
    }

    @JsonProperty
    @Override
    public Instant getTime() {
        return time;
    }

    @Override
    public String toString() {
        return StringUtils.allFieldsToString(this);
    }

}

File: src/main/kotlin/com/flexpoker/game/command/events/BaseGameEvent.kt
package com.flexpoker.game.command.events

import com.fasterxml.jackson.annotation.JsonProperty
import com.flexpoker.exception.FlexPokerException
import com.flexpoker.game.command.framework.GameEvent
import com.flexpoker.util.StringUtils
import java.time.Instant
import java.util.UUID

/**
 * Base class used to handle some standard methods and fields so that the
 * subclasses can just be specific to what they need.
 */
abstract class BaseGameEvent(private val aggregateId: UUID) : GameEvent {
    private var version = 0
    private val time: Instant
    @JsonProperty("gameId")
    override fun getAggregateId(): UUID {
        return aggregateId
    }

    @JsonProperty
    override fun getVersion(): Int {
        if (version == 0) {
            throw FlexPokerException("should be calling getVersion() in situations where it's already been set")
        }
        return version
    }

    override fun setVersion(version: Int) {
        this.version = version
    }

    @JsonProperty
    override fun getTime(): Instant {
        return time
    }

    override fun toString(): String {
        return StringUtils.allFieldsToString(this)
    }

    init {
        time = Instant.now()
    }
}